### ISSUE TITLE
Is the automake build using clang toolchain currently broken?

### ISSUE THREAD

**Issue by jmftrindade** [memkind/memkind#340]
Hi there!

I tried to build memkind with clang toolchain instead of gcc (required for our project) earlier today, and got a number of build errors.  In the end I managed to build it (see below for details) after disabling a number of build warnings.  Some of these, however, look a little suspicious (e.g., `Wtautological-compare`, and `Wsign-compare` which is explicitly enabled in the g++ build).  Here are the steps you can follow to reproduce:

Env:
```
clang v5.0.0
g++ v7.3.1

# uname -a
Linux JF1121-080309T 4.15.6-300.fc27.x86_64 #1 SMP Mon Feb 26 18:43:03 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

# cat /etc/fedora-release
Fedora release 27 (Twenty Seven)
```

Build steps:
```bash
export CC=clang
export CXX=clang++
./autogen.sh && ./configure
make
```

To get the clang build to work, I had to disable the warnings below in the following order:

**Wtautological-compare**
```
(...)
  CC       src/memkind.lo
src/memkind.c:274:16: error: comparison of unsigned enum expression >= 0 is always true
      [-Werror,-Wtautological-compare]
    if((policy >= 0) && (policy < MEMKIND_POLICY_MAX_VALUE)) return 0;
(...)
```

**Wunused-parameter**
```
(...)
  CC       src/memkind_arena.lo
src/memkind_arena.c:185:42: error: unused parameter 'extent_hooks'
      [-Werror,-Wunused-parameter]
void *arena_extent_alloc(extent_hooks_t *extent_hooks,
(...)
 CXX      examples/memkind_cpp_allocator.o
In file included from examples/memkind_cpp_allocator.cpp:34:
./include/memkind_allocator.h:193:50: error: unused parameter 'n'
      [-Werror,-Wunused-parameter]
            void deallocate(pointer p, size_type n) const
(...)
```

**Wsign-compare**
```
(...)
  CC       src/memkind_hbw.lo
src/memkind_hbw.c:318:36: error: comparison of integers of different signs: 'unsigned long'
      and 'int' [-Werror,-Wsign-compare]
        assert(hbw_node_mask->size >= nodes_num);
(...)
```

In the end I also had to change the autogen'd LIBS, CXXFLAGS, and CFLAGS to the following:
```bash
make "CXXFLAGS=-Wall -Wextra -Wno-unused-parameter -g3 -fvisibility=hidden -O3 -fPIC -std=c++11" "CFLAGS=-Wall -Wextra -Wundef -Wno-format-zero-length -Wno-tautological-compare -Wno-unused-parameter -Wno-sign-compare -pipe -g3 -fvisibility=hidden -O3 -funroll-loops -fPIC" "LIBS=-lm -pthread -ldl -lnuma"
```

These are quite different from the ones generated by autogen + configure, e.g., -fPIC was missing, -std=gnu11 had to be removed, etc.

Build seems to work fine for our project now, but the default values generated by configure seem to be currently broken for clang.  Please advise, thanks!

**Comment by michalbiesek** [comment]
Hi,
Thanks for your message.
Regarding support of clang toolchain - we test the build in Travis with Clang-8.
Regarding warnings, which you reported part of them is fixed by #341.
For short term solution, you could remove Werror flag from your build, just like it was done in #51 
For long term solution, I think we implement a solution proposed in https://github.com/memkind/memkind/pull/51#issuecomment-567951787

**Comment by jmftrindade** [comment]
Thanks for the prompt response @michalbiesek!  I can confirm that latest `master` branch builds as expected for clang toolchain on clang v5.0 (instead of Travis's clang v3.8, aka Clang-8).
